### DISCUSSION TITLE
Testers

### DISCUSSION THREAD

**Discussion by arneschreuder** [COS-301/graduates#14] [General]
This is the discussion board for and with Testers.

**Comment by Alexander-Muendesi** [comment]
@COS-301/testers Good day

As seen from our other discussion board for Testers, we do we have to come up with a test plan, formulate where tests are to be created, document for developers what tests should look like and give some examples in this first Sprint.

Since this has to be agreed upon by all testers a meeting is required to come up with something. Potential suggestions of when to hold this meeting are welcome

**Reply by Joseph-Makgopa** [reply]
I suggest 11/03/2020 in the afternoon.

**Reply by jadepeche** [reply]
If we do it on 11/03/2020, it'll have to be before 5 because I have a team meeting at 5.

**Reply by JohanCSmit** [reply]
Can we please do it a few days from now to allow time for research and preparation.

**Reply by Alexander-Muendesi** [reply]
I agree with Johan. Tomorrow is a bit too short notice and the meeting will likely achieve nothing as I dont think people have prepared enough/done enough research

**Reply by jademichellepeche** [reply]
Okay I agree with that. Would Monday (14/03/2022) be okay?

**Reply by Alexander-Muendesi** [reply]
I am personally casting my vote for Saturday 19 March simply because :
1) A lot of us have COS332 Semester test 1 next week
2) Cos301 class test next week

I feel like people will be busy preparing and stuff and at least it would give people enough time to research on what Sir told us to do

**Reply by Joseph-Makgopa** [reply]
Do you mean the 12th of March?

**Reply by Joseph-Makgopa** [reply]
When is the meeting?

**Reply by Alexander-Muendesi** [reply]
If you check the Discord Server for 301, people have voted for this coming Monday . Time is still being voted for but it seems 1430 is the winner currently

**Comment by Alexander-Muendesi** [comment]
@COS-301/project-managers Please take note of this conversation which was held in a discussion for testers not visible to everyone
![image](https://user-images.githubusercontent.com/93581519/157689776-da8b0445-4fba-4197-80ae-457693969dcb.png)
![image](https://user-images.githubusercontent.com/93581519/157689870-ab52cd4d-fe1e-4689-b051-78dd0776691c.png)
![image](https://user-images.githubusercontent.com/93581519/157690084-fa2c3b0d-bcb0-49a8-be7c-9ea0814708e4.png)

**Comment by Joseph-Makgopa** [comment]
Good day @arneschreuder 

I am a tester for the JavaScript team working on the PWA feature. Our business analyst is still working on the sub features So in the meantime I was doing research on the sub features that I think will need testing and I also have a few questions below.


Testing PWA
    • Validate PWA Manifest(name, start URL property set, different size icons, display property set)
    • Validate Service worker
    • Verify if website is served via HTTPS
    • Verify if website loads offline
    • Verify if web pages are responsive
    • Verify if Metadata for Home Screen is provided.
    • Verify if Page transitions are consistent on all network speeds
    • Testing Push notifications


I have the following questions regarding testing.

    • How can we show we tested some of the core features of a PWA such as responsiveness (support of mobile and desktop views) and multiple browser compatibility which require us to use a device emulator or manually test the app across different browsers. They can not be automated to produce results which can be shown.
    • Some features such as the app being installable require manual verification either on a browser or a device and can not be logged for evaluation.
    • Some features require manual verification using browser tools.
    • Is the testing of the PWA features restricted to the provided technologies or can we suggest other tools if we find that the provided testing technologies can not test other features.

I recognize that these questions are implementation focused. I just wanted some clarity before writing the documentation so that I am not describing testing procedures that might make it difficult for the person implementing them in the second sprint to present to the Project owners as there might not be any  actual deliverables (e.g. scripts and logs) to present or maybe my role should not exist for this feature.

**Comment by Vhulenda88** [comment]
Do we as testers also use git actions or is it just the CI/CD team

**Reply by arneschreuder** [reply]
Not directly no, but indirectly yes. So for example. Testers need to know which tests are executed, how, how to look for failures, how to report on issues found. So for example, lets say my PR failed, and this was because of some failed unit test. I expect the DevOps role, the PM, myself to fix it and the tester should assist here needed. After the test is fixed, that feature needs to be tested in isolation for usability. This is totally outside of CI/CD realm unless you are doing some wild e2e testing etc.

**Comment by Joseph-Makgopa** [comment]
Good day @arneschreuder 

In terms of testing for the pwa feature. Is it correct to say that the tester needs to write (describe) tests that check that the other features' requirements are working in a certain way when offline or is that the responsibility of the other teams to write offline requirements and tests for offline functionality.If the former is the case does that mean if there are 'x' combined functional requirements from other teams then we have tests for 'x' functional requirements in an offline context if possible. 

An example I would give is that Progressive web apps such as twitter and youtube keep you temporarily logged In even when offline, for our system is this the responsibility of the authorization team or PWA team to test that it keeps you logged In when offline.

**Reply by arneschreuder** [reply]
Hi There. I think I already answered this in some other discussion. But you should not see it so clearly categories and organised. There is a LOT of overlap between teams and roles. The PWA team is ultimately responsible for EVERYTHING Mobile + PWA related. HOWEVER, it is NOT their responsibility to implement a feature to be offline available. These remain the responsibilities of the feature teams. BUT, how would a feature team KNOW their feature needs to be available offline? the PWA/Mobile team needs to drive that feature, if implementation help is needed, then so be it. For example, the offline capability and then authentication team (note, not authorization) go hand in hand, since how do you know you are offline? This is the PWA+Mobile teams implementation responsibility, and auth might use that. The PWA+Mobile team might need to provide libraries to make offline integration easier (cache etc.) So for example, to get offline right with Angular + GraphQL, you can use GraphQL's cache, but that is used EVERYWHERE by UI Engineers. But they need to know to use cache otherwise offline is not a concern for them. So this is where the PWA+Mobile team comes in and says, ok, these features need to be usable offline, lets make that happen. Work with the relevant feature teams to get done what must be done.


The testers DO NOT WRITE tests. They just tests. Usability testing.

**Comment by Vhulenda88** [comment]
Where should we keep the test logs and are we supposed to have our own separate testing plans for each feature or do we have just one testing plan for the whole project

**Reply by Vhulenda88** [reply]
@COS-301/testers

**Comment by Vhulenda88** [comment]
@COS-301/testers  just to clarify, we can only do e2e testing after everyone has merged all their PRs to develop right.... or can we do feature specific e2e testing ?

**Comment by JonathanEnslin** [comment]
@COS-301/testers I would like to make a few changes/clarifications on the testing wiki regarding logging tests. It is as follows:

- Add the PR number to the log for future logs, e.g. PR: 264
- Clarify whether developers and testers real name should be used in the log
- Come up with a naming convention for the logs, I am thinking, if we make a log for PR 91 for example, we call the file PR91.txt, assuming we store the logs as text files, if we use the PR number in the filename we could leave it out of the log, although I feel it would be a good idea to still include it

Then we also need to discuss where on the repo we will keep the logs, because last time I checked we have not come up with a place yet.

Can someone comment on whether they think I can add these changes or if it's unnecessary.

**Comment by JonathanEnslin** [comment]
Good day @arneschreuder 
A while ago you said in this reply to a discussion - https://github.com/COS-301/graduates/discussions/14#discussioncomment-2413458 - that testers to do not write tests, so I have only been reviewing tests and logging the results. But should the testers still write e2e tests, and when doing the usability and acceptance testing, should that be done together with the BAs?

**Reply by JonathanEnslin** [reply]
@arneschreuder is it possible we could get clarification on this?

**Comment by KearonF** [comment]
For integration and e2e tests, has anyone managed to serve the api, db and client together for cypress testing? In a way that it can be automated, because I can get it running locally by manually starting the different components.

**Reply by KearonF** [reply]
@Alexander-Muendesi, would we perhaps have to add serving the api and db to the CI/CD pipeline when doing the automated tests as to allow for integration and e2e tests to utilise them?

**Reply by arneschreuder** [reply]
Yeah, you kinda need an environment to run those tests against. You have 2 options:

- Instantiate an environment in the CI/CD pipeline. This pipeline will have its own environments etc.
- Or talk to the delivery team to get a development env deployed, and deploy there and run e2e-tests post deployment in dev.

**Reply by KearonF** [reply]
Thank you for your response.

@Alexander-Muendesi, please discuss above solutions with your team.

**Reply by Alexander-Muendesi** [reply]
it is mentioned CI/CD or Delivery team @KearonF . Dont necessarily shove it to us especially on a last minute basis like this. This should have been brought to our attention way earlier. We can make no guarantees as we already have a bunch of other work we need to sort out

**Reply by KearonF** [reply]
@JonathanEnslin please see above discussion and discuss above possible solutions with your team, and discuss changes to the ci/cd with @Alexander-Muendesi for either proposed solution

**Reply by JonathanEnslin** [reply]
@arneschreuder, will a testing environment for e2e tests have to be set up in a docker container if we do not integrate it in the CI/CD pipeline? I do not see any methods for setting it up on github other than using the CI/CD pipeline, and the delivery team only recently got access to the server that the docker images will be hosted on.

**Comment by KearonF** [comment]
Good Morning @arneschreuder, as there is currently no test environment implemented in which we can run integration and e2e tests automatically, should we merge testing PRs after extensive local testing? This does however raise the issue of breaking the CI/CD pipeline. 

Please advise possible alternatives if that is not acceptable.

**Comment by t3ddy33** [comment]
Good afternoon, @arneschreuder. The rest of the testers and myself are unsure of where exactly to place the testing logs. Is there any location in particular in which they should be stored?
